src/lib/swapper/swappers/OsmosisSwapper/endpoints.ts

```diff
@@ -0,0 +1,210 @@
+import {
+  buildApiTradeTx,
+  buildPerformIbcTransferUnsignedTx,
+  symbolDenomMapping,
+} from 'lib/swapper/swappers/OsmosisSwapper/utils/helpers'
+import { createDefaultStatusResponse } from 'lib/utils/evm'
+
+import { getTradeQuote } from './getTradeQuote/getTradeQuote'
+import { atomOnOsmosisAssetId, COSMO_OSMO_CHANNEL } from './utils/constants'
+import type { OsmosisSupportedChainId } from './utils/types'
+
+const tradeQuoteMetadata: Map<string, { chainId: OsmosisSupportedChainId }> = new Map()
+
+export const osmosisApi: Swapper2Api = {
+  getTradeQuote: async (
+    input: GetTradeQuoteInput,
+    { sellAssetUsdRate }: { sellAssetUsdRate: string },
+  ): Promise<Result<TradeQuote2, SwapErrorRight>> => {
+    const tradeQuoteResult = await getTradeQuote(input, { sellAssetUsdRate })
+
+    return tradeQuoteResult.map(tradeQuote => {
+      const { receiveAddress, affiliateBps } = input
+      const id = uuid()
+      // TODO(gomes): getTradeQuote() is currently a hack, which represents the swap as a single trade, and needs to be revamped
+      // This effectively means that getUnsignedTx() won't be able to be implemented for all steps, but only the first for now
+      // i.e either the IBC transfer or swap-exact-amount-in
+      tradeQuoteMetadata.set(id, {
+        chainId: tradeQuote.steps[0].sellAsset.chainId as OsmosisSupportedChainId,
+      })
+      return { id, receiveAddress, affiliateBps, ...tradeQuote }
+    })
+  },
+
+  getUnsignedTx: async ({
+    from,
+    tradeQuote,
+    stepIndex,
+  }: GetUnsignedTxArgs): Promise<CosmosSignTx> => {
+    const { accountNumber, buyAsset, sellAsset, sellAmountBeforeFeesCryptoBaseUnit } =
+      tradeQuote.steps[stepIndex]
+    const { receiveAddress } = tradeQuote
+
+    const sellAssetIsOnOsmosisNetwork = sellAsset.chainId === osmosisChainId
+
+    const sellAssetDenom = symbolDenomMapping[sellAsset.symbol as keyof SymbolDenomMapping]
+    const buyAssetDenom = symbolDenomMapping[buyAsset.symbol as keyof SymbolDenomMapping]
+
+    const adapterManager = getChainAdapterManager()
+    const osmosisAdapter = adapterManager.get(osmosisChainId) as osmosis.ChainAdapter | undefined
+    const cosmosAdapter = adapterManager.get(cosmosChainId) as cosmos.ChainAdapter | undefined
+
+    const { REACT_APP_OSMOSIS_NODE_URL: osmoUrl } = getConfig()
+
+    if (!cosmosAdapter || !osmosisAdapter) throw new Error('Failed to get adapters')
+
+    let sellAddress
+
+    if (sellAssetIsOnOsmosisNetwork) {
+      sellAddress = from
+
+      if (!sellAddress)
+        throw new SwapError('failed to get osmoAddress', {
+          code: SwapErrorType.EXECUTE_TRADE_FAILED,
+        })
+    } else {
+      /** If the sell asset is not on the Osmosis network, we need to bridge the
+       * asset to the Osmosis network first in order to perform a swap on Osmosis DEX.
+       */
+      sellAddress = from
+
+      if (!sellAddress) throw new Error('Failed to get address')
```

Comment:
```suggestion
    if (!from) throw new Error('Missing from address')

    const sellAddress = from

    /** If the sell asset is not on the Osmosis network, we need to bridge the
     * asset to the Osmosis network first in order to perform a swap on Osmosis DEX.
     */
    if (!sellAssetIsOnOsmosisNetwork) {
```

Reply:
Agree this is turbo ugly but going to keep it like this in this PR, this is just a trimmed version of the current implementation and there is actual refactor of it coming as a direct follow-up, keeping it as close as possible to the original will make things easier here

Reply:
Cleaned up in https://github.com/shapeshift/web/pull/4905